dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/DefaultPredictionService.java

```diff
@@ -449,186 +360,44 @@ public void predict( Predictor predictor, Date startDate, Date endDate, Predicti
-     * <p>
-     * This is needed because we may get multiple data values that need to be
-     * aggregated to the same item value. In the case of a
-     * DimensionalItemObject, this may be multiple values from children
-     * organisation units. In the case of a DataElement, this may be multiple
-     * value from children organisation units and/or it may be multiple
-     * disaggregated values that need to be summed for the data element.
-     * <p>
-     * Note that a single data value may contribute to a DataElementOperand
-     * value, a DataElement value, or both.
-     *
-     * @param item the item to add (the DataElementOperand or DataElement).
-     * @param items the set of items to be added (only if item is in set).
-     * @param dv the data value to be added.
-     * @param value the numeric value to be added.
-     * @param aocData the datavalue map by attribute option combo to add to.
-     */
-    private void addToData( DimensionalItemObject item, Set<DimensionalItemObject> items,
-        DataValue dv, Object value, MapMapMap<String, Period, DimensionalItemObject, Object> aocData )
-    {
-        if ( !items.contains( item ) )
-        {
-            return;
-        }
-
-        Object valueSoFar = aocData.getValue( dv.getAttributeOptionCombo().getUid(), dv.getPeriod(), item );
-
-        Object valueToStore = (valueSoFar == null) ? value : addDoubleObjects( value, valueSoFar );
-
-        aocData.putEntry( dv.getAttributeOptionCombo().getUid(), dv.getPeriod(), item, valueToStore );
-    }
-
-    /**
-     * For a predictor and orgUnit, determines the set of attribute option
-     * combos for which predictions will be generated.
-     *
-     * @param itemMap item data map for an orgUnit.
-     * @param defaultCategoryOptionCombo system default category option combo.
-     * @return set of attribute option combos to use for an orgUnit.
-     */
-    Set<String> getAttributeOptionCombos(
-        MapMapMap<String, Period, DimensionalItemObject, Object> itemMap,
-        CategoryOptionCombo defaultCategoryOptionCombo )
-    {
-        Set<String> attributeOptionCombos = new HashSet<>( itemMap.keySet() );
-
-        if ( attributeOptionCombos.isEmpty() )
-        {
-            attributeOptionCombos.add( defaultCategoryOptionCombo.getUid() );
-        }
-
-        return attributeOptionCombos;
-    }
-
-    /**
-     * Checks to see if a dimensional item object has values stored in the
-     * database by attribute option combo.
-     *
-     * @param o dimensional item object
-     * @return true if values are stored by attribuete option combo.
-     */
-    private boolean hasAttributeOptions( DimensionalItemObject o )
-    {
-        return o.getDimensionItemType() != DimensionItemType.PROGRAM_INDICATOR
-            || ((ProgramIndicator) o).getAnalyticsType() != AnalyticsType.ENROLLMENT;
+        return new HashSet<>( Sets.union( analyticsQueryPeriods, existingOutputPeriods ) );
```

Comment:
No, it's not mutated further. I wanted a clean `HashSet` for debugging and performance. But that probably wasn't a good choice for two reasons: (a) it would raise doubts about mutability just like the one you had, and (b) unlike some of the other data that is being initialized and used many, many times, this is actually used only once since there's just one `datavalue` query. I've removed the `HashSet` creation and am just returning the `union`.